ethcore/light/src/on_demand/response_guard.rs

```diff
@@ -0,0 +1,184 @@
+	Decoder,
+	/// Empty response
+	EmptyResponse,
+	/// Wrong header sequence
+	HeaderByNumber,
+	/// Too few results
+	TooFewResults,
+	/// Too many results
+	TooManyResults,
+	/// Trie error
+	Trie,
+	/// Unresolved header
+	UnresolvedHeader,
+	/// No responses expected.
+	Unexpected,
+	/// Wrong hash
+	WrongHash,
+	/// Wrong Header sequence
+	WrongHeaderSequence,
+	/// Wrong response kind
+	WrongKind,
+	/// Wrong number
+	WrongNumber,
+	/// Wrong Trie Root
+	WrongTrieRoot,
+}
+
+/// Handle and register responses that can fail
+#[derive(Debug)]
+pub struct ResponseGuard {
+	state: failsafe::StateMachine<ResponsePolicy, ()>,
+	responses: HashMap<Inner, usize>,
+}
+
+impl ResponseGuard {
+	/// Constructor
+	pub fn new(window_dur: Duration) -> Self {
+		let policy = failsafe::failure_policy::success_rate_over_time_window(1.0, 1, window_dur, NoBackoff);
+		Self {
+			state: failsafe::StateMachine::new(policy, ()),
+			responses: HashMap::new(),
+		}
+	}
+
+	fn into_reason(&self, err: &ResponseError<super::request::Error>) -> Inner {
+		match err {
+			ResponseError::Unexpected => Inner::Unexpected,
+			ResponseError::Validity(ValidityError::BadProof) => Inner::BadProof,
+			ResponseError::Validity(ValidityError::Decoder(_)) => Inner::Decoder,
+			ResponseError::Validity(ValidityError::Empty) => Inner::EmptyResponse,
+			ResponseError::Validity(ValidityError::HeaderByNumber) => Inner::HeaderByNumber,
+			ResponseError::Validity(ValidityError::TooFewResults(_, _)) => Inner::TooFewResults,
+			ResponseError::Validity(ValidityError::TooManyResults(_, _)) => Inner::TooManyResults,
+			ResponseError::Validity(ValidityError::Trie(_)) => Inner::Trie,
+			ResponseError::Validity(ValidityError::UnresolvedHeader(_)) => Inner::UnresolvedHeader,
+			ResponseError::Validity(ValidityError::WrongHash(_, _)) => Inner::WrongHash,
+			ResponseError::Validity(ValidityError::WrongHeaderSequence) => Inner::WrongHeaderSequence,
+			ResponseError::Validity(ValidityError::WrongKind) => Inner::WrongKind,
+			ResponseError::Validity(ValidityError::WrongNumber(_, _)) => Inner::WrongNumber,
+			ResponseError::Validity(ValidityError::WrongTrieRoot(_, _)) => Inner::WrongTrieRoot,
+		}
+	}
+
+	/// Update the state after a `faulty` call
+	pub fn register_error(&mut self, err: &ResponseError<super::request::Error>) -> Result<(), Error> {
+		self.state.on_error();
+		let err = self.into_reason(err);
+		*self.responses.entry(err).or_insert(0) += 1;
+		if self.state.is_call_permitted() {
+			trace!(target: "circuit_breaker", "ResponseGuard: {:?}, responses {:?}", self.state, self.responses);
+			Ok(())
+		} else {
+			trace!(target: "circuit_breaker", "ResponseGuard: {:?}, responses: {:?}", self.state, self.responses);
+			let (&err, &max_count) = self.responses.iter().max_by_key(|(_k, v)| *v).expect("got at least one element; qed");
+			let majority = self.responses.values().filter(|v| **v == max_count).count() == 1;
+			// FIXME: more efficient with a separate counter
```

Comment:
Would be nice if you can create an new issue and include an issue number there!

Reply:
I will fix this if/when @cheme signs off of this :P

Reply:
Sorry to ask, did I lock something?

Reply:
@cheme 

haha, I was referring to the functionality that this PR breaks i.e, `replacing the response limit per peer` and to use `a simple timeout` instead? If, that is not sorted there is no point and implementing other things IMHO!

Reply:
Ok, thanks for taking the time but to verify that I understood your concerns:

1. This PR still might be useful (I want to add that it gives some information to the user why the request to the full-node failed or least what the majority response was)
2. Circuit breaker for `responses` are useless and we should replace it with a simple timeout as stated you pointed out earlier thanks for that!
3. Exponentially backoff for requests is useful but because we only register `bad responses` here to it is better to use `consecutive failures` (also easiser to understand)

>  I just observe that there is a hard limit at 30 attempts for exponential backoff), here I think that using consecutive failure strategy could make sense.

I think you are wrong there as you can see, https://github.com/dmexe/failsafe-rs/blob/master/src/backoff.rs#L136 the backoff is always regenerated. I think the `MAX_RETRIES` is used to calcuate "binary exponential backoff" by shifting the value `1 <<  MAX_RETRIES` and bigger value than 63 will overflow. That's my best guess 

> Still, I do not really get why we use success rate over consecutive failure (success rate can only be 0 or 100 (request end on first success and we use our constructions on a per request basis) whereas consecutive failure takes a lot of possible states and is somehow managed as the error counter in request_guards (I may be wrong I did not read the crate implementation)).

Basically, the request will run `forever` until it gets a successful response in the `circuit breaker`. That is why we use an upperbound  `max_failures` in the RequestGuard when to drop the request

It is okay for me to try with `consecutive failures instead` but then I suggest setting to value to 1 and then backoff and repeat until we either got a successful response or backoff #max_number in order to avoid doing too many N/W requests.

Reply:
> I think you are wrong there as you can see, https://github.com/dmexe/failsafe-rs/blob/master/src/backoff.rs#L136 the backoff is always regenerated. I think the MAX_RETRIES is used to calcuate "binary exponential backoff" by shifting the value 1 << MAX_RETRIES and bigger value than 63 will overflow. 

Oups, sorry, I should have check more cautiously, thanks for looking at it.

> Basically, the request will run forever until it gets a successful response in the circuit breaker. That is why we use an upperbound max_failures in the RequestGuard when to drop the request

That is what I understood :-)

I feel like consecutive failure before backoff could make sense as a percentage of the current number of peers, but I do not really have things to back that (except that when testing a few month back I usually got a lot of peers returning invalid value but when we do not get lot of peers we want to wait). Isn't it a kind of error rates strategy. I start to wonder if it may make sense to write our own tailored 'failure policy' like 'ErrorRateOverTimeWindow' (just thinking loud).

I do not know about dropping response_guard, am I correct if I say that 'request_guard' and 'response_guard' are both in 'Pending' struct ? (just wondering if they can be merged, it probably only make sense with custom 'failure policy')

Reply:
> I do not know about dropping response_guard, am I correct if I say that 'request_guard' and 'response_guard' are both in 'Pending' struct ? (just wondering if they can be merged, it probably only make sense with custom 'failure policy')

Yes, they are. Thus the first of `response_guard` and `request_guard` to detect a failure will drop the `Pending`. The idea from the beginning was to have a two failure policies of the `circuit-breaker` and combine them but because the state can't be accessed and it will never terminate I created two wrappers instead. The problem is that we can just query if the request is permitted or not thus can't tell if it will backoff will be closed/half-open or stay open forever. So, we can either an additional timeout or keep some state by our selves. 

I don't really understand what you are suggesting and the relationship between `requests` and `responses` when you talk about `our own faliure policy`.

What I can pick off is this:
- We should start with consecutive failures based on the number of peers
- When the peers are under some threshold we should backoff?
 - ?

Well, I think the circuit breaker API doesn't provide the functinality to change the peers for example after the circuit breaker has been created which makes that solution `hard` to get right (unless we want to write an circuit breaker ourselves)

Personally, I don't have any inutution on what this values should be and suggest just to keep it simple (e.g, consecutive failures then backoff and continue backoff and so on). The consecutive failures shall be configured by the user

Reply:
:+1: for keeping it simple, I don't either have strong intuition on value but I identify two case where my previous configuration where not good : 
- start of a light client : there is to few nodes connected and we probably want to wait
- well established connection in light client : there will be no new node so waiting does not make a lot of sense.

About  'our own failure policy', I was mainly interpolating that our need may be  some 'RateOverTimeWindow', but as we count error and not success our own implementation of failsafe trait `FailurePolicy` could be an idea. Thinking twice, going for the crate success rate counting could be another target, but it would involve managing circuit breaker at a per peer level instead of a per request level. So I see two scenario (I do not say that they should be include in this pr):
- we will keep per request management, so we count errors only -> go with error count policy or a custom error rate one.
- we plan to switch to per peerid circuit so we can count multiple success (having mutex in the circuit does make sense with this scenario) and keeping a rate success is explain by this future move. The issue here is that hour request handler are not threadsafe (not sure they need to be, maybe they can stay per request specific but using per user id circuit breakers).

Reply:
All right, yeah I think it is out-of-scope for this PR but we can create an issue and investigate it further!